node-graph/gcore/src/graphic_element.rs

```diff
@@ -38,6 +41,67 @@ impl AlphaBlending {
 	}
 }
 
+#[derive(Clone, Debug, Default, serde::Serialize, serde::Deserialize)]
+pub struct Instances<T>
+where
+	T: Into<GraphicElement> + StaticType + 'static,
+{
+	id: Vec<InstanceId>,
+	instances: Vec<Arc<Mutex<T>>>,
+}
+
+impl<T: Into<GraphicElement> + StaticType + 'static> Instances<T> {
+	pub fn new(instance: T) -> Self {
+		Self {
+			id: vec![InstanceId::generate()],
+			instances: vec![Arc::new(Mutex::new(instance))],
+		}
+	}
+
+	pub fn instances(&self) -> impl Iterator<Item = MutexGuard<'_, T>> {
+		self.instances.iter().map(|item| item.lock().expect("Failed to lock mutex"))
+	}
+
+	pub fn id(&self) -> impl Iterator<Item = InstanceId> + '_ {
+		self.id.iter().copied()
+	}
+
+	pub fn push(&mut self, id: InstanceId, instance: T) {
+		self.id.push(id);
+		self.instances.push(Arc::new(Mutex::new(instance)));
+	}
+
+	pub fn replace_all(&mut self, id: InstanceId, instance: T) {
+		let mut instance = Arc::new(Mutex::new(instance));
+
+		for (old_id, old_instance) in self.id.iter_mut().zip(self.instances.iter_mut()) {
+			let mut new_id = id;
+			std::mem::swap(old_id, &mut new_id);
+			std::mem::swap(&mut instance, old_instance);
+		}
+	}
+}
+
+impl<T: Into<GraphicElement> + Hash + StaticType + 'static> core::hash::Hash for Instances<T> {
+	fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
+		self.id.hash(state);
+		for instance in &self.instances {
+			let instance = instance.lock().unwrap();
+			instance.hash(state);
+		}
+	}
+}
+
+impl<T: Into<GraphicElement> + PartialEq + StaticType + 'static> PartialEq for Instances<T> {
+	fn eq(&self, other: &Self) -> bool {
+		self.id == other.id && self.instances.len() == other.instances.len() && { self.instances.iter().zip(other.instances.iter()).all(|(a, b)| *a.lock().unwrap() == *b.lock().unwrap()) }
```

Comment:
I'm pretty sure it will deadlock if you compare an array to a clone of itself. This is because the same mutexes will be in self and other. Therefore when we try and get a lock on both a and b at the same time which can never happen.